### PR TITLE
gui: Kernel options page

### PR DESCRIPTION
**Pull request by evil-brain**
Adds a kernel options page to the GUI. Plugs into the back end in a similar manner to the same page on the TUI.
Has a feature for default kernels based on what's loaded in the model already, this means virtualbox installations will by default choose the correct kernel.

Removing more than one flag is problematic, it seems that we just pass the whole value to CBM and it doesn't split strings on space and remove all args.

Fixes Issue: #384, #406

Changes proposed in this pull request:
- Add an advanced menu option to modify kernel-related settings
- Page contains a list of all available kernels, highlights and selects the default, adds text to the default name to indicate default status.
- Localization support is added, but still needs to be properly reviewed and properly vetted.

### CODE REVIEW

File: gui/pages/kernel_options.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+// GetID returns the ID for this page
+func (page *ConfigKernelPage) GetID() int {
+	return PageIDConfigKernel
+}
+
+// GetIcon returns the icon for this page
+func (page *ConfigKernelPage) GetIcon() string {
+	return "computer"
```
Comment (mdhorn) [inline gui/pages/kernel_options.go]:
```suggestion
	return "applications-engineering"
```

Comment (mdhorn) [reply gui/pages/kernel_options.go]: Please update to a unique icon; I like this one best for now.

File: locale/en_US/LC_MESSAGES/clr-installer.po [hunk 2]
```diff
@@ -543,6 +543,33 @@ msgstr "Installation in progress."
...
+msgid "Add Extra Arguments:"
+msgstr "Add Extra Arguments:"
+
+msgid "Remove Arguments:"
+msgstr "Remove Arguments:"
+
+msgid "Kernel Configuration"
+msgstr "Kernel Configuration"
+
+msgid ", custom command line arguments"
```
Comment (mdhorn) [inline locale/en_US/LC_MESSAGES/clr-installer.po]: Please drop the comma from the string. Instead, let's have two sentences.

File: locale/en_US/LC_MESSAGES/clr-installer.po [hunk 3]
```diff
@@ -543,6 +543,33 @@ msgstr "Installation in progress."
 msgid "Installation successful."
 msgstr "Installation successful."
 
+msgid " (default)"
+msgstr " (default)"
+
+msgid "Add Extra Arguments:"
+msgstr "Add Extra Arguments:"
```
Comment (mdhorn) [inline locale/en_US/LC_MESSAGES/clr-installer.po]: Please drop the : from the label strings to be consistent with the User Add screen

File: gui/pages/kernel_options.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+		}
+		box.PackStart(labelDesc, true, true, 0)
+
+		page.list.Add(box)
+	}
+
+	// addLabel: Tell users they can add kernel command lines below
+	addText := utils.Locale.Get("Add Extra Arguments:")
+
+	page.addLabel, err = setLabel(addText, "label-addLabel", 0.0)
```
Comment (mdhorn) [inline gui/pages/kernel_options.go]: Please change the style to the "label-entry" which we are using for the add user screen for consistency.

File: gui/pages/kernel_options.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+		return nil, err
+	}
+	page.addEntry.SetMarginStart(common.StartEndMargin)
+	page.addEntry.SetMarginEnd(common.StartEndMargin)
+	page.addEntry.SetTooltipText(utils.Locale.Get(kernelArgsHelp))
+	page.box.PackStart(page.addEntry, false, false, 0)
+
+	// remLabel label
+	remText := utils.Locale.Get("Remove Arguments:")
+	page.remLabel, err = setLabel(remText, "label-remLabel", 0.0)
```
Comment (mdhorn) [inline gui/pages/kernel_options.go]: Please change the style to the "label-entry" which we are using for the add user screen for consistency.

File: gui/pages/kernel_options.go [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+	}
+	if _, err := page.list.Connect("row-activated", page.onRowActivated); err != nil {
+		return nil, err
+	}
+	page.scroll.Add(page.list)
+
+	// Create list data
+	for _, v := range page.data {
+		name := v.Name
+		desc := v.Desc
```
Comment (mdhorn) [inline gui/pages/kernel_options.go]: Please add a call to the util localization to the description of the kernel and add entries to the po files.
The name may not make sense to translate, but if we leave that in English having the description translated would be helpful.

Comment (evil-brain) [reply gui/pages/kernel_options.go]: Done.

File: gui/pages/kernel_options.go (hunk 5 above)
Comment (evil-brain) [reply gui/pages/kernel_options.go]: Done.

File: gui/pages/kernel_options.go (hunk 4 above)
Comment (evil-brain) [reply gui/pages/kernel_options.go]: Done.

File: locale/en_US/LC_MESSAGES/clr-installer.po (hunk 3 above)
Comment (evil-brain) [reply locale/en_US/LC_MESSAGES/clr-installer.po]: Done.

File: locale/en_US/LC_MESSAGES/clr-installer.po (hunk 2 above)
Comment (evil-brain) [reply locale/en_US/LC_MESSAGES/clr-installer.po]: Done

File: gui/pages/kernel_options.go (hunk 1 above)
Comment (evil-brain) [reply gui/pages/kernel_options.go]: Done.

**Comment by evil-brain** [comment]
> I can't seem to clear the kernel arguments. I would expect if I clear the field and Confirm that it would be removed. Also, if I clear and put something new, it is added to the previous value.

Fixed this in the next push.

Also, I guess I should add a massive banner to my motd or maybe a commit hook:
```
    ____  __  ___   __   ________  ________   __    _____   __________________  ______
   / __ \/ / / / | / /  /_  __/ / / / ____/  / /   /  _/ | / /_  __/ ____/ __ \/ / / /
  / /_/ / / / /  |/ /    / / / /_/ / __/    / /    / //  |/ / / / / __/ / /_/ / / / / 
 / _, _/ /_/ / /|  /    / / / __  / /___   / /____/ // /|  / / / / /___/ _, _/_/_/_/  
/_/ |_|\____/_/ |_/    /_/ /_/ /_/_____/  /_____/___/_/ |_/ /_/ /_____/_/ |_(_|_|_)   

```

**Comment by mdhorn** [comment]
> There is a small bug right now where if you click the "Confirm" button on the kernel screen with no values for the command line it crashes. I've emailed you a patch; which will also fix the text-based user interface.

In case the email didn't work, here is a gist
https://gist.github.com/mdhorn/10155812faf1464df9857c124359f349

**Comment by evil-brain** [comment]
Applied/tested Mark's patch, it fixes the issue for me! Thanks, Mark!

File: gui/pages/kernel_options.go [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+				name = v.Name + utils.Locale.Get(" (default)")
+			}
+		}
+
+		box, err := setBox(gtk.ORIENTATION_VERTICAL, 0, "box-list-label")
+		if err != nil {
+			return nil, err
+		}
+
+		labelName, err := setLabel(name, "list-label-name", 0.0)
```
Comment (reaganlo) [inline gui/pages/kernel_options.go]: I don't think this style "list-label-name" is defined.

File: gui/pages/kernel_options.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+		if err != nil {
+			return nil, err
+		}
+		box.PackStart(labelName, false, false, 0)
+
+		labelDesc, err := setLabel(desc, "list-label-desc", 0.0)
```
Comment (reaganlo) [inline gui/pages/kernel_options.go]: I don't think this style "list-label-desc" is defined.

File: gui/pages/kernel_options.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	// Box
+	page.box, err = setBox(gtk.ORIENTATION_VERTICAL, 0, "box-page")
+	if err != nil {
+		return nil, err
+	}
+
+	// ScrolledWindow
+	page.scroll, err = setScrolledWindow(gtk.POLICY_NEVER, gtk.POLICY_AUTOMATIC, "scroller")
```
Comment (reaganlo) [inline gui/pages/kernel_options.go]: Can we have a heading for this? Like "Select Kernel"

File: gui/pages/kernel_options.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+	page.addLabel, err = setLabel(addText, "label-entry", 0.0)
+	if err != nil {
+		return nil, err
+	}
+	page.addLabel.SetMarginStart(common.StartEndMargin)
+	page.addLabel.SetHAlign(gtk.ALIGN_START)
+	page.box.PackStart(page.addLabel, false, false, 10)
+
+	// addEntry: Args to add to the kernel command line
+	page.addEntry, err = setEntry("entry")
```
Comment (reaganlo) [inline gui/pages/kernel_options.go]: This style "entry" is adding an extra top padding causing additional space between heading and the entry. Can we remove this additional space by defining another style for this?

File: gui/pages/kernel_options.go [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+	page.remLabel, err = setLabel(remText, "label-entry", 0.0)
+	if err != nil {
+		return nil, err
+	}
+	page.remLabel.SetMarginStart(common.StartEndMargin)
+	page.remLabel.SetMarginEnd(common.StartEndMargin)
+	page.box.PackStart(page.remLabel, false, false, 10)
+
+	// remEntry: Args to remove to the kernel command line.
+	page.remEntry, err = setEntry("entry")
```
Comment (reaganlo) [inline gui/pages/kernel_options.go]: This style "entry" is adding an extra top padding causing additional space between heading and the entry. Can we remove this additional space by defining another style for this?

**Comment by evil-brain** [comment]
Thanks for the patch @reaganlo!

At this point, all 3 of us worked on this, I think we can almost call this "crowd sourced source"  :stuck_out_tongue_winking_eye:

**Pull request merged**